### ISSUE TITLE
Quaternion norm and real_field not in args

### ISSUE THREAD

**Issue by evbernardes** [sympy/sympy#24943]
In `Quaternion` an extra argument called `norm` to allow the definition of a quaternion with a pre-determined norm. This is useful to define symbolic unit quaternions that simplify a lot of symbolic computation. Defining:
```python
from sympy import Quaternion
from sympy.abc import a, b, c, d
```
`Quaternion(a, b, c, d).norm()` returns `sqrt(a**2 + b**2 + c**2 + d**2)`, while `Quaternion(a, b, c, d, norm=1).norm()` returns `1`.

However, `norm` isn't added in the `args`, so making that:
```python
q1 = Quaternion(a, b, c, d, norm=1)
q2 = q1.func(*q1.args).norm()
``` 
Gives  `sqrt(a**2 + b**2 + c**2 + d**2)` instead of `1`.

Is that something that should be fixed?

Moreover, the argument `real_field` has the same problem.

**Comment by oscarbenjamin** [comment]
All data defining the object should be included as part of args. I think that there was a previous attempt to define a UnitQuaternion class that would have unit norm. I don't remember where that ended up.

I think that the real_field argument is a bad idea in general.

**Comment by evbernardes** [comment]
> All data defining the object should be included as part of args. I think that there was a previous attempt to define a UnitQuaternion class that would have unit norm. I don't remember where that ended up.
> 
> I think that the real_field argument is a bad idea in general

What do you think if I remove the option to set the norm after the quaternion has been created, and then add the pre-set norm to args?

Also not sure what to do with `real_field`, so I'll leave that alone for now.

**Comment by evbernardes** [comment]
Alternatively, a `UnitQuaternion` class that simply inherits everything from `Quaternion` but changes the definition of `.norm()` should be easy to implement, since the methods from `Quaternion` all take into account internal calls to `norm` anyways

**Comment by daijapan** [comment]
Enjoy reading this issue. If you have pure math problem - let us solve it. Or is this purely coding problem?

**Comment by kostas-rigan** [comment]
@evbernardes I am currently working on defining a `UnitQuaternion` class, and I have a question. In case a `UnitQuaternion` is created with numerical values, should an error be raised or should the arguments be normalized? For example:
```python
>>> uq = UnitQuaternion(1, 2, 3, 4)
>>> uq.norm()
1
```
If  `Quaternion(1, 2, 3, 4)` was defined then `norm` would return `sqrt(30)`.
```python
>>> q = Quaternion(1, 2, 3, 4)
>>> q.norm()
sqrt(30)
```
The normalized arguments would be:
```python
>>> q.normalize().args
(sqrt(30)/30, sqrt(30)/15, sqrt(30)/10, 2*sqrt(30)/15)
```

**Comment by evbernardes** [comment]
> @evbernardes I am currently working on defining a `UnitQuaternion` class, and I have a question. In case a `UnitQuaternion` is created with numerical values, should an error be raised or should the arguments be normalized? For example:

Hey there @kostas-rigan, depends on what the others think, but personally I think an error is best, to avoid doing something the user isn't expecting. 

I recently implement a `norm` argument in the `Quaternion` class that I can set in order for me to force the quaternion's norm to a pre-defined value (in this case, it can be 1, for example.)

You can study it to see the solution I came up with: if all the quaternions's elements evaluate to a number, then I test if the actual calculated norm is compatible with the pre-set norm. Example:
```python
>>> q = Quaternion(1, 2, 3, 4, norm=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../sympy/sympy/algebras/quaternion.py", line 122, in __new__
    obj.set_norm(norm)
  File ".../sympy/sympy/algebras/quaternion.py", line 158, in set_norm
    _check_norm(self.args, norm)
  File ".../sympy/sympy/algebras/quaternion.py", line 27, in _check_norm
    raise ValueError("Incompatible value for norm.")
ValueError: Incompatible value for norm.
```

Special care should also be taken when a value is substituted, example:
```python
>>> q = Quaternion(a, 2, 3, 4, norm=1)
>>> q.subs(a, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../sympy/sympy/core/basic.py", line 1040, in subs
    rv = rv._subs(old, new, **kwargs)
  File "...sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File ".../sympy/sympy/core/basic.py", line 1152, in _subs
    rv = self._eval_subs(old, new)
  File ".../sympy/sympy/algebras/quaternion.py", line 1013, in _eval_subs
    _check_norm(elements, norm)
  File ".../sympy/sympy/algebras/quaternion.py", line 27, in _check_norm
    raise ValueError("Incompatible value for norm.")
ValueError: Incompatible value for norm.
>>> 
```

I think for the substitution case, normalizing without saying anything is bad because I don't think it makes sense, and it would be hiding something from the user.

On the other hand, implementing a normalization on `__init__` and an error in the `subs` is inconsistent.

So, from a user's point of view, I'd prefer to always have errors, to let me know this quaternion I'm supposing has norm 1 doesn't actually.

**Comment by evbernardes** [comment]
Moreover, this would mostly be used with non-numerical values anyways, so I think this minor inconvenience of having an error would hardly be an actual inconvenience during practical use, while it could be useful in order to quickly find errors in the reasoning.

**Comment by kostas-rigan** [comment]
I understand, thank you for the explanation! I will try to cover both cases.

**Comment by evbernardes** [comment]
Good luck! In any case, try not to recode everything by hand. Inheritance is probably the way to go here: by creating a `UnitQuaternion` class that inherits from `Quaternion`, you probably won't have to code much apart from changing the `norm` function.

Moreover, this:
```python
q = UnitQuaternion()
isinstance(q, Quaternion)
```

Would be true, which I think is desirable.

**Comment by kostas-rigan** [comment]
Hello, I've defined the `UnitQuaternion` class inheriting from `Quaternion` as we've discussed. I created a `__new__` method which simply returns a `Quartenion` object with `norm=1` in the arguments. 
```python
def __new__(cls, a=0, b=0, c=0, d=0, real_field=True):
        return Quaternion.__new__(cls, a, b, c, d, real_field=real_field, norm=1)
```
I also created the `norm` method returning 1 to the user. I believe these were the only two methods needed to be redefined for UnitQuaternion. Of course if there is anything else to be done, I can look into it!

**Comment by evbernardes** [comment]
If you're using the `norm` argument as defined in the `Quaternion` class, you don't need to redefine the `norm()` method.

The idea of having a `UnitQuaternion` class, though, would be to avoid the use of this `norm` argument since it's not a part of `args`.

Moreover, some special care should be taken in the multiplication in relation of the classes returned after each operation.

Example: `1 * UnitQuaternion` returns a `UnitQuaternion`, but `2 * UnitQuaternion` returns a quaternion.

Moreover, `UnitQuaternion * UnitQuaternion` -> `UnitQuaternion` but `UnitQuaternion * Quaternion` -> `Quaternion`

**Comment by evbernardes** [comment]
> I understand! My concern was that internally `_norm` would be None, therefore if someone substituted the values to numbers that didn't match norm=1 it wouldn't inform the user about incompatible values

To solve this problem, I included a check inside the implementation of `_eval_subs` (this is the method that has to be changed if something special is expected when doing substitutions):

```python
    def _eval_subs(self, *args):
        elements = [i.subs(*args) for i in self.args]
        norm = self._norm
        try:
            norm = norm.subs(*args)
        except AttributeError:
            pass
        _check_norm(elements, norm)
        return Quaternion(*elements, norm=norm)
```

> Maybe by calling `set_norm(1)` inside the **new** method, would resolve it without providing a `norm` argument. 

The problem with this is that it still relies on this `norm` argument from the original `Quaternion` class. This is either problematic or it isn't:

- If this is problematic, then it is to be expected that this will be removed later. Having another class that depends on this would complicate things.
- If this is not problematic, then I'm not sure creating a `UnitQuaternion` class has much advantage, since it wouldn't propagate pre-defined norms as well.

Other than multiplication, is there any other operation, like division for example, that needs special care?

Division, definitely. Not sure about others though, maybe check if exponentiation by `1` returns a `UnitQuaternion`.

**Comment by oscarbenjamin** [comment]
The previous attempt to introduce a UnitQuaternion class is gh-18776.

It is not clear that there is any real advantage in having a UnitQuaternion class. I also think that the norm argument is misguided. Whatever problem it solves should be solved in some other way for example if it is known in a calculation that the quaternions should be unit quaternions then there should be no need to normalise them.

**Comment by evbernardes** [comment]
The absolute best would be to have some sort of system that you can add constraints to variables. I suppose it would be a monumental task to implement this, though.

@oscarbenjamin do you think adding the norm argument to `args` would break things? I could potentially add it, run the tests and see what would happen.

Alternatively, maybe I could add some kind of "normed quaternion" class that has this norm argument in `args`

**Comment by oscarbenjamin** [comment]
I think that it is better to take a step back and consider why anyone would want any of these features.

What problem does the norm argument solve in the first place? Why not just remove it completely?

If the purpose of the norm argument is to avoid dividing by an expression for a norm that is known to be 1 then why not just avoid dividing by the norm altogether?

**Comment by oscarbenjamin** [comment]
> The previous attempt to introduce a UnitQuaternion class is [gh-18776](https://github.com/sympy/sympy/pull/18776).

CC @mgeier

**Comment by evbernardes** [comment]
> I think that it is better to take a step back and consider why anyone would want any of these features.

I have used the `Quaternion` class extensively during my thesis to demonstrate many general results that I needed to demonstrate using quaternions with symbolic elements.

> What problem does the norm argument solve in the first place? Why not just remove it completely?

Many of the different methods in quaternion use an internal call to `norm`. For example, `to_rotation_matrix`, which I suppose is widely used.

> If the purpose of the norm argument is to avoid dividing by an expression for a norm that is known to be 1 then why not just avoid dividing by the norm altogether?

Since I was dealing with unit rotation quaternions, in my tests, in order to use these methods and do my analyses, I had to multiple/divide many times by the norm in order to counter these internal calls to norm, which was very frustrating and prone to errors.

Basically, I did this implementation for myself in which I just told the class to suppose the quaternion has a fixed norm and then it got a LOT easier to keep dealing with equations. I agree that this implementation how I did might not be the prettiest solution for now, but I proposed it to sympy because, at least in practice, it helped me a lot for doing some more esoteric quaternionic analysis.

**Comment by oscarbenjamin** [comment]
> Many of the different methods in quaternion use an internal call to `norm`. For example, `to_rotation_matrix`, which I suppose is widely used.

Maybe that should be changed. Perhaps the `to_rotation_matrix` method could have a parameter to prevent dividing by the norm if the quaternion can be assumed to be a unit quaternion. Perhaps that should be the default behaviour even since it is only meaningful to identify unit quaternions with rotation matrices anyway. The docstring could just say "this method assumes that the quaternion is a unit quaternion".

**Comment by evbernardes** [comment]
> Maybe that should be changed. Perhaps the `to_rotation_matrix` method could have a parameter to prevent dividing by the norm if the quaternion can be assumed to be a unit quaternion. Perhaps that should be the default behaviour even since it is only meaningful to identify unit quaternions with rotation matrices anyway. The docstring could just say "this method assumes that the quaternion is a unit quaternion".

If you construct a quaternion from numerical elements that weren't properly normalized, though, it wouldn't produce a proper rotation matrix. On the top of my head, I can think of this problem, but I'd argue it might cause other bugs.

**Comment by evbernardes** [comment]
In this case I think a `UnitQuaternion` class is a good idea: one if you're expecting a completely general quaternion, and another one if you're expecting a unit quaternion.

Yes, rotations are well expressed by unit quaternions. But not everything is. Two examples:

1.  the quaternion form of the angular velocity for example is a non-unit quaternion, used, for example, to integrate rotation quaternions:
```math
\frac{dq}{dt} = \frac{1}{2} q \odot \hat{\omega}
```
Where:

```math
\hat{\omega} = \begin{bmatrix} 0 \\ \omega \end{bmatrix}
```

2. Dual quaternions, used to represent both rotations and displacements, are defined with two different quaternions. One is a unit-quaternion, defining the rotation, the other one is non-unit.

If we're making assumptions that the quaternions is unit, then I think having a separate `UnitQuaternion` class that doesn't depend on this norm argument would be the clearer way of expressing it, I think this would be easier than having separate methods.

**Comment by oscarbenjamin** [comment]
> rotations are well expressed by unit quaternions. But not everything is.

Yes, but those other quaternions should not be used with `to_rotation_matrix`. If `to_rotation_matrix` should only ever be called with a unit quaternion then why should it divide by the norm?

**Comment by evbernardes** [comment]
I think the problem is that it would go against conventions coming from maths. When using a quaternion to represent a rotation, we use the conjugation operation that is [defined](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation) with the inverse as:

```math
p' = q \odot p \odot q^{-1} \text{\;\;\;\;(1)}
```

Instead of the conjugate:

```math
p' = q \odot p \odot q^{*} \text{\;\;\;\;(2)}
```

By removing the squared norm from the definition of the rotation matrix, you're effectively implementing equation 2, which should just be a simplification. By applying the norm, `to_rotation_matrix` would always return a rotation matrix.

Btw, I'm sorry if I don't have your experience with Sympy and best practices. I'm by no means a developer, which is why I'm trying to contribute with my point of view of someone working with engineering/applied mathematics and trying to use Sympy. Supposing the norm is always 1 would be adding some sort of hidden assumption and would make the whole implementation less useful to do anything that isn't very classical. Moreover I find it to be confusing, since it's inconsistent with every other implementation I've seen of quaternions I've seen.

**Comment by evbernardes** [comment]
Here's an example of a real application. Suppose I want to implement in my robot a little code that takes a quaternion and transforms it into Euler angles. I could do this:
```python
>>> from sympy import Quaternion
>>> from sympy.abc import a, b, c, d
>>> q = Quaternion(a, b, c, d)
>>> q.to_euler('ZYX', avoid_square_root=True)
```
And get this equation for the second angle, which would deal with non-normalized quaternions (useful in practice):
```math
\theta_2 = \text{asin}{\left(\frac{- \left(a - c\right)^{2} + \left(a + c\right)^{2} + \left(- b + d\right)^{2} - \left(b + d\right)^{2}}{2 a^{2} + 2 b^{2} + 2 c^{2} + 2 d^{2}} \right)}
```

Now if my implementation has a normalization step beforehand, I can do this instead:
```python
>>> q = Quaternion(a, b, c, d, norm=1)
>>> q.to_euler('ZYX', avoid_square_root=True)
```

And, with a simple call to `simplify`, get this:
```math
\theta_2 = - \text{asin}{\left(2 a c - 2 b d \right)}
```

Removing these internal calls for normalization would remove a lot of the simplicity in using Sympy to deduce solutions some actual application.

**Comment by evbernardes** [comment]
And well, it would also be an API break with people currently already using `Quaternion`.

**Comment by oscarbenjamin** [comment]
> I think the problem is that it would go against conventions coming from maths

There is no mathematical convention for using non-unit quaternions for rotations. That applies to `to_rotation_matrix` and `to_euler` and anything else that tries to interpret a quaternion as representing a rotation. The two formulae that you have shown are equivalent for unit quaternions and rotations are only defined for unit quaternions so either formula can be used.

> which would deal with non-normalized quaternions (useful in practice)

Why can't the user just call `normalize` if that is what is intended? Why should SymPy invent a non-existent convention for what it means to associate Euler angles with a non-unit quaternion (especially if that complicates the formulas in the more important case where a unit quaternion is used)?

> And well, it would also be an API break with people currently already using `Quaternion`.

Exactly what should happen by default or how to manage changing things is a separate question but firstly we should consider what the ideal behaviour should be.

**Comment by evbernardes** [comment]
> firstly we should consider what the ideal behaviour should be

Well, I think we just won't agree on this one, so I'll stop here. 

Whatever you decide to do, though, feel to ping me, so I can help implement the necessary changes.

**Comment by mgeier** [comment]
For reference, these are my previous PRs about unit quaternions: #17694, #18065, #18086, #18776.

And there was another one later, not by me, but by @dhyams: #21902

The problem with my own approaches was that I wasn't able to do anything non-trivial with unit quaternions, see the discussion starting at https://github.com/sympy/sympy/pull/18776#issuecomment-787466433 for a still extremely simple task that couldn't be done.
As long as this task cannot be done, it IMHO doesn't make sense to have unit quaternions at all.

I would be curious though: @evbernardes you seem to have used SymPy's `Quaternion` class (or some extension thereof?) ... do you happen to have an example for anything useful that you have done with it?

The only thing I have ever used the `Quaternion` class for was a derivation of a certain convention of converting Euler angles to quaternions: https://audioscenedescriptionformat.readthedocs.io/en/latest/quaternions.html. This was helpful, but it was still quite trivial. I guess I could have also done that by manually implementing a function for quaternion multiplication with 4-tuples.

**Comment by evbernardes** [comment]
> I would be curious though: @evbernardes you seem to have used SymPy's Quaternion class (or some extension thereof?) ... do you happen to have an example for anything useful that you have done with it?

@mgeier I have used Sympy's `Quaternion` class a lot, indeed!

One of the things I did was to verify formulas I derived for the general conversion from a quaternion to Euler angles in any sequence. I ended up finding some pretty nice formulas that became a [publication](https://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0276302) and I then implemented it both in Scipy and Sympy.

> The only thing I have ever used the Quaternion class for was a derivation of a certain convention of converting Euler angles to quaternions

You can now use `Quaternion.to_euler` and `Quaternion.from_euler` for that :)

Another thing I did was to work in a new type of decomposition of quaternion rotations that I used in a specific control architecture, which is currently under review in another publication.

**Comment by evbernardes** [comment]
> The problem with my own approaches was that I wasn't able to do anything non-trivial with unit quaternions, see the discussion starting at https://github.com/sympy/sympy/pull/18776#issuecomment-787466433 for a still extremely simple task that couldn't be done.
As long as this task cannot be done, it IMHO doesn't make sense to have unit quaternions at all.

For this, I think the only way would be for Sympy to have some way of giving assumptions/constraints between variables. Using the `norm` argument like I implemented now, I have:
```python
>>> from sympy import Quaternion, symbols
>>> q1 = Quaternion(*symbols('{w:z}'), norm=1)
>>> q2 = q1 * q1.inverse()
```
This gives:
```python
>>> q1
{w} + {x}*i + {y}*j + {z}*k
>>> q2
({w}**2 + {x}**2 + {y}**2 + {z}**2) + 0*i + 0*j + 0*k
```
And (I think) there isn't a way of showing Sympy that `({w}**2 + {x}**2 + {y}**2 + {z}**2) = 1`. But by propagating the norms when multiplying quaternions, I have:
```python
>>> q1.norm()
1
>>> q2.norm()
1
```
Which I personally find useful, but I agree it's still not amazing.

**Comment by evbernardes** [comment]
@mgeier One possibly very inefficient idea to improve that would be to add a call to `replace` every time a Quaternion product is done. Adding this to `Quaternion._general_mul`:
```python
if q1._norm is not None:
    q = q.replace(q1.a ** 2 + q1.b ** 2 + q1.c ** 2 + q1.d ** 2, q1._norm)
if q2._norm is not None:
    q = q.replace(q2.a ** 2 + q2.b ** 2 + q2.c ** 2 + q2.d ** 2, q2._norm)
```
Now I have:
```
>>> from sympy import Quaternion, symbols [ from sympy import Quaternion, symbols] 
>>> q1 = Quaternion(*symbols('{w:z}'), norm=1)
>>> q1 * q1.inverse()
1 + 0*i + 0*j + 0*k
```
Which is maybe a hint of what could be done. Maybe have a `UnitQuaternion` class that does automatic calls to `subs`, `replace`, `factor`, etc. in an intelligente way?

**Comment by oscarbenjamin** [comment]
> And (I think) there isn't a way of showing Sympy that `({w}**2 + {x}**2 + {y}**2 + {z}**2) = 1`

There is a way of simplifying an expression subject to a polynomial relation like this:
```python
In [2]: w, x, y, z = symbols('w, x, y, z')

In [3]: q1 = Quaternion(w, x, y, z)

In [4]: one = q1 * q1.inverse()

In [5]: one
Out[5]: 
(w**2/(w**2 + x**2 + y**2 + z**2) + x**2/(w**2 + x**2 + y**2 + z**2) + y**2/(w**2 + x**2 + y**2 + z
**2) + z**2/(w**2 + x**2 + y**2 + z**2)) + 0*i + 0*j + 0*k

In [6]: one.args[0]
Out[6]: 
         2                   2                   2                   2       
        w                   x                   y                   z        
───────────────── + ───────────────── + ───────────────── + ─────────────────
 2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2
w  + x  + y  + z    w  + x  + y  + z    w  + x  + y  + z    w  + x  + y  + z 

In [7]: ratsimpmodprime(one.args[0], [q1.norm()**2 - 1])
Out[7]: 1
```
Of course other simplification methods could easily handle this particular case but `ratsimpmodprime` is quite general. If Quaternion was not implemented as a separate class but using `Add` and `Mul` then it would be possible to apply ratsimpmodprime directly to the quaternion rather than using `.args[0]`.

Of course it is better to avoid creating complicated expressions in the first place rather than relying on post-simplifying them later.

**Comment by evbernardes** [comment]
I think the problem would most be keep track of all the different expressions that should be simplified to `1`.

@oscarbenjamin what do you think if a `UnitQuaternion` class had some kind of list that keeps track of all of the different norms that should be simplified to 1?

For example, defining:
```python
q0 = Quaternion(*symbols('{w:z}^0'))
q1 = Quaternion(*symbols('{w:z}^1'))
q2 = Quaternion(*symbols('{w:z}^2'))
```

And a list that stores everything that should be simplified (this list could be updated every time two `UnitQuaternion` instances are multiplied):
```python
L = [q0.norm()**2, q1.norm()**2, q2.norm()**2]
```

Then doing some (dirty atm) simplification:
```python
def simp(q):
  q_simplified = q
  for l in L:
       q_simplified = q_simplified.factor(l)
       q_simplified = q_simplified.replace(l, 1)
  return q_simplified
```

Which could be called everytime the UnitQuaternion is multiplied with some other UnitQuaternion. This would result in:
```python
>>> qa = simp(simp(q2 * q1.inverse()) * simp(q1 * q0.inverse()))
>>> qb = simp(q2 * q0.inverse())
>>> simp(qa, L) == simp(qb, L)
True
```

**Comment by oscarbenjamin** [comment]
I think that automatic simplification is usually a bad approach. It is much better to either:

1. Avoid creating something that might need simplification in the first place.
2. Have an explicit simplification function.

If you have an explicit simplification function then you can pass in whatever information it needs. Following point 1 means avoiding norm in the first place wherever it is redundant e.g. by using `conjugate` instead of `inverse` if you know that you have a unit quaternion.

**Comment by mgeier** [comment]
By "explicit simplification function" do you mean something that is still implicitly called by `sympy.simplify()`?

Or do you mean that the end user would need to know which specific simplification function to call?

In my original example code (https://github.com/sympy/sympy/pull/18776#issuecomment-787472428) I've already used `sp.simplify(x1 - x2)`. Would that be enough?

> e.g. by using `conjugate` instead of `inverse` if you know that you have a unit quaternion.

That's how I implemented `inverse()` in #18776.

**Comment by evbernardes** [comment]
> I think that automatic simplification is usually a bad approach. It is much better to either:
>     1. Avoid creating something that might need simplification in the first place.

I think I see why it can be a bad approach. Using, e.g., `conjugate` instead `inverse` like you said is what I do when I construct a proof using Sympy from beginning to end.

But in practice, having this unit norm assumption was pretty useful for me when I wanted to directly test a final result I got after some pages of tedious handwritten algebra.

Btw when I said:
 > there isn't a way of showing Sympy that `w**2 + x**2 + y**2 + z**2 = 1`

I meant it in an automatic manner, like a way of telling Sympy assumptions to try to simplify everytime an operation is done, something like:

1. Save result of operation as `expr`
2. Check if `expr` contains atoms `(a, b)`
3. If it does, try to simplify it with `ratsimpmodprime(expr, [a**2 + b**2 - 1])`

It could be considered a bad approach, like you said, but I would probably use it extensively if it existed.

**Comment by oscarbenjamin** [comment]
> By "explicit simplification function" do you mean something that is still implicitly called by `sympy.simplify()`?
> 
> Or do you mean that the end user would need to know which specific simplification function to call?

I think that the user somehow needs to specify that the quaternion is a unit quaternion so they should call a function that allows them to specify that (unlike simplify).

> Using, e.g., `conjugate` instead `inverse` like you said is what I do when I construct a proof using Sympy

It's also worth considering what you would do with pen and paper. Would you divide by the norm if you knew it was a unit quaternion or would you just automatically use the conjugate formula? Why would you expect sympy to have to divide by a norm and then apply an algorithm to simplify the norm away if that is not how you would otherwise do the calculation?

> I meant it in an automatic manner, like a way of telling Sympy assumptions to try to simplify **everytime** any operation is done

This is bad for many reasons:

1. It makes everything slow because now you have no way to stop simplification from happening **everytime** you do **anything** and good simplification can be very slow even when it does not do anything useful.
2. There are usually reasons for wanting to choose between different ways of representing something and automatic simplification makes it impossible because the expression always converts to a different form e.g.:
```python
>>> print(1/sqrt(2))
sqrt(2)/2
```
3. Automatic simplification can only ever do one kind of simplification. There is no way to pass options or arguments or anything to say exactly what you do or don't want to be simplified or what assumptions should be made.
4. Generally it is better for users to understand simplification as an explicit process because then they have some control and some way of understanding what is happening because they can see its effect in isolation. It also makes everything more understandable because they can read the docs for the simplification function or they can see if the simplification is the part that is slow.

**Comment by evbernardes** [comment]
I meant it as something that should be explicitly enabled by the user, not something by default. In the sense that the users themselves would add each of the simplifications.

I agree that otherwise this would just be slow and useless.

**Comment by evbernardes** [comment]
@sylee957 I don't think I understood what you proposed as a change, did you make a PR about this already?

**Comment by sylee957** [comment]
I would ask a specific part you find hard to understand,
However, I would list some concrete reasons why it should be changed, and the steps about changing it

- First of all, assigning to object argument in SymPy is breaking the rule of SymPy
- Second of all, being able to mutate the definitions of SymPy objects is not correct. so either `set_norm` should be discouraged or deprecated. Even outside of mathematical programming, I rarely find mutable objects useful at all, because I always suspect about when and how it can change its value.

So anyway,
- The first step I would propose is that `norm, real-field` should be part of sympified `Quaternion.args`, or either stay as immutable field by comments, docstring, or type annotations, such that people don't get misguided about them.
- The second step I would propose is to deprecate `set_norm`.

And now it remains the third step, we need some alternative about how `norm` used to do, 
and how we need an alternative solution as much as it does not violate the rules in SymPy.

I previously done an attempt to throw out https://github.com/sympy/sympy/pull/25734, 
but for that part, throwing out the `_a, b, _c, _d` was very beautiful.
But throwing out `norm, real_field` may not be easy because it needs to deal with deprecation issues.

**Comment by evbernardes** [comment]
Indeed I agree with making it unmutable and that we can remove the `norm` thing, as it seems to have created more hard than good.

I meant that I didn't understand what changed when you said this:
> It does not suffer from the limitations of SymPy simplify or assumptions not able to handle the algebraic identities like sqrt(a**2 + b**2 + c**2 + d**2) == 1

And about the squared norm, you are proposing to store the norm as a square and then return the square root, or are you proposing that the function `norm` should compute and return the square norm directly?

**Comment by evbernardes** [comment]
> The reason we try to keep backward compatibility and alternatives for even such broken rules, is very optional.

Given how recent this is, maybe it wouldn't be a problem to remove it completely though, don't you think?
